javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -135,14 +133,11 @@ private Lazy(Supplier<? extends T> supplier) {
      */
     @Override
     public T get() {
-        // using a local var speeds up the double-check idiom by 25%, see Effective Java, Item 71
-        Supplier<? extends T> tmp = supplier;
-        if (tmp != null) {
+        if ((value == null) && (supplier != null)) {
```

Comment:
I think we have a lot bigger problem, than optimizations: all the tests are passing with non-volatile fields also!

I tried creating a test that would fail, but haven't come up with one yet (only if I put 
`Try.run(() -> Thread.sleep(ThreadLocalRandom.current().nextInt(2)))` all over the production code):
```java
@Test
public void shouldBeConsistentFromMultipleThreads() throws Exception {
    final Vector<Tuple2<Lazy<Integer>, Integer>> values = Vector.range(0, 10000).map(i -> {
        final Integer expected = ((i % 2) == 1) ? null : i;
        final AtomicBoolean isEvaluated = new AtomicBoolean(false);
        final Lazy<Integer> lazy = Lazy.of(() -> {
            Try.run(() -> Thread.sleep(ThreadLocalRandom.current().nextInt(2)));
            assertThat(isEvaluated.getAndSet(true)).isFalse();
            return expected;
        });
        return Tuple.of(lazy, expected);
    });

    CompletableFuture.allOf(
            values.flatMap(t -> range(0, 10)
                    .map(j -> runAsync(() -> assertThat(t._1.get()).isEqualTo(t._2)))
            ).toJavaArray(CompletableFuture.class)
    ).join();
}
```

@danieldietrich, @jbgi, @zsolt-donca, can someone tell me why this isn't failing for non-volatile fields, i.e.:
```java
private Supplier<? extends T> supplier;
private T value;

@Override
public T get() {
    if (value == null) {
        synchronized (this) {
            if (supplier != null) {
                value = supplier.get();
                supplier = null;
            }
        }
    }
    return value;
}
```

(it IS failing if we do a non-synchronized, single check)